File: benchit/Benchit-V5/kernel/utilities/skeleton/C/0/PCL/0/kernel_main.c
/******************************************************************************
*
*  B e n c h I T - Performance Measurement for Scientific Applications
*
*  c pcl kernel skeleton
*  This file: interface between the kernel and BenchIT.
*
*  Author: Robert Wloch (wloch@zhr.tu-dresden.de)
*  Last change by: $Author: mickler $
*  $Revision: 1.6 $
*  $Date: 2005/12/13 00:40:33 $
*
******************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include "interface.h"
/*  Header for local functions
 */
#include "work.h"

/** These variables will help us to keep the overview over the arrays
  * we access for our functions/data.
  */
/* Number of different ways an algorithm will be measured.
   Example: loop orders: ijk, ikj, jki, jik, kij, kji -> n_of_works=6 with
   each different loop order in an own function. */
int n_of_works;
/* Number of fixed functions we have per measurement.
   Example: execution time and MFLOPS are measured for each loop order
   -> n_of_sure_funcs_per_work=2 */
int n_of_sure_funcs_per_work;

int MIN, MAX, INCREMENT;

#ifdef BENCHIT_USE_PCL
int *wish_list;
int wish_nevents;
int *doable_list;
int doable_nevents;
unsigned int mode;
PCL_DESCR_TYPE descr;
#endif

/*  Header for local functions
 */
void evaluate_environment( void );

/**  The implementation of the bi_getinfo from the BenchIT interface.
 *   Here the infostruct is filled with informations about the
 *   kernel.
 *   @param infostruct  a pointer to a structure filled with zero's
 */
void bi_getinfo( bi_info * infostruct )
{
   int i = 0, j = 0; /* loop var for n_of_works */
#ifdef BENCHIT_USE_PCL
   int status = -1;
   mode = PCL_MODE_USER;
   /* B ########################################################*/
   wish_nevents = 1;
   /*########################################################*/
   wish_list = (int *)malloc( wish_nevents * sizeof( int ) );
   if ( wish_list == 0 )
   {
      fprintf( stderr, "Allocation of wish_list failed.\n"); fflush( stderr );
      exit( 127 );
   }
   /* B ########################################################*/
   wish_list[0] = PCL_INSTR;
   /*########################################################*/
   if( PCLinit( &descr ) != PCL_SUCCESS )
   {
      fprintf( stderr, "Unable to initialize PCL.\n" ); fflush( stderr );
      exit( 1 );
   }

   if (DEBUGLEVEL > 0) {
      /* the next four lines are only for your information and can be ereased later */
      printf("\n*******************************************************************************\n");
      PCLstatus(descr);
      printf("*******************************************************************************\n");
      fflush(stdout);
   }

   status = PCLcheck(descr, wish_list, wish_nevents, &doable_list, &doable_nevents, mode);
   if (status != PCL_SUCCESS) {
      printf("None of your PCL Events can be counted on this machine!\n"); fflush(stdout);
      exit(127);
   }

   if (DEBUGLEVEL > 1) {
      /* the next two for loops: */
      /* this is for your information only and can be ereased if the kernel works fine */
      for (i=0; i<wish_nevents; i++) {
         int j = 0, hit = 0;
         printf("wish_list event %d (%s) is ", i, PCLeventname(wish_list[i]));
         for (j=0; j<doable_nevents; j++) {
            if (wish_list[i] == doable_list[j]) {
               hit=1;
               break;
            }
         }
         if (hit == 0) printf("not ");
         printf("measurable\n");
      }
      for (i=0; i<doable_nevents; i++) {
         printf("doable_list event %d (%s)\n", i, PCLeventname(doable_list[i]));
      }
   }

   /* just in case, although it's really not neccessary after PCLcheck */
   if (PCLquery(descr, doable_list, doable_nevents, mode) != PCL_SUCCESS){
      printf("PCL event(s) not measurable\n"); fflush(stdout);
      exit(1);
   }
#endif
   (void) memset (infostruct, 0, sizeof (bi_info));
   /* get environment variables for the kernel */
   evaluate_environment();
   /* B ########################################################*/
   infostruct->codesequence = bi_strdup("work_*()");
   infostruct->xaxistext = bi_strdup( "Problem Size" );
   infostruct->maxproblemsize = (MAX-MIN+1)/INCREMENT;
   if( (MAX-MIN+1) % INCREMENT != 0 )
     infostruct->maxproblemsize++;
   infostruct->num_processes = 1;
   infostruct->num_threads_per_process = 0;

   infostruct->kernel_execs_mpi1 = 0;
   infostruct->kernel_execs_mpi2 = 0;
   infostruct->kernel_execs_pvm = 0;
   infostruct->kernel_execs_omp = 0;
   infostruct->kernel_execs_pthreads = 0;
   n_of_works = 2; /* number versions of this algorithm (ijk, ikj, kij, ... = 6 */
   n_of_sure_funcs_per_work = 2; /* time measurement and MFLOPS (calculated) */
   /*########################################################*/
#ifdef BENCHIT_USE_PCL
      infostruct->numfunctions = n_of_works*(n_of_sure_funcs_per_work + doable_nevents);
#else
      infostruct->numfunctions = n_of_works*n_of_sure_funcs_per_work;
#endif

   /* allocating memory for y axis texts and properties */
   infostruct->yaxistexts = malloc( infostruct->numfunctions * sizeof( char* ) );
   if ( infostruct->yaxistexts == 0 )
   {
     fprintf( stderr, "Allocation of yaxistexts failed.\n" ); fflush( stderr );
     exit( 127 );
   }
   infostruct->outlier_direction_upwards = malloc( infostruct->numfunctions * sizeof( int ) );
   if ( infostruct->outlier_direction_upwards == 0 )
   {
     fprintf( stderr, "Allocation of outlier direction failed.\n" ); fflush( stderr );
     exit( 127 );
   }
   infostruct->legendtexts = malloc( infostruct->numfunctions * sizeof( char* ) );
   if ( infostruct->legendtexts == 0 )
   {
     fprintf( stderr, "Allocation of legendtexts failed.\n" ); fflush( stderr );
     exit( 127 );
   }
   infostruct->base_yaxis = malloc( infostruct->numfunctions * sizeof( double ) );
   if ( infostruct->base_yaxis == 0 )
   {
     fprintf( stderr, "Allocation of base yaxis failed.\n" ); fflush( stderr );
     exit( 127 );
   }
   /* setting up y axis texts and properties */
   for ( j = 0; j < n_of_works; j++ )
   {
      /* B ########################################################*/
      int index1 = 0 * n_of_works + j;
      int index2 = 1 * n_of_works + j;
      //int index3 = 2 * n_of_works + j;
      // 1st function
      infostruct->yaxistexts[index1] = bi_strdup( "Calculation Time in s" );
      infostruct->outlier_direction_upwards[index1] = 1;
      infostruct->base_yaxis[index1] = 0;
      // 2nd function
      infostruct->yaxistexts[index2] = bi_strdup( "FLOPS" );
      infostruct->outlier_direction_upwards[index2] = 0;
      infostruct->base_yaxis[index2] = 0;
      /*########################################################*/
      // 3rd function
      //infostruct->yaxistexts[index3] = bi_strdup( "" );
      //infostruct->outlier_direction_upwards[index3] = 0;
      //infostruct->base_yaxis[index3] = 0;
#ifdef BENCHIT_USE_PCL
      for ( i = 0; i < doable_nevents; i++ )
      {
         char buf[180];
         int index = n_of_works * ( n_of_sure_funcs_per_work + i ) + j;
         if ( doable_list[i] < PCL_MFLOPS )
            sprintf( buf, "%160s per second", PCLeventname( doable_list[i] ) );
         else
            sprintf( buf, "%160s", PCLeventname( doable_list[i] ) );
         infostruct->yaxistexts[index] = bi_strdup( buf );
         infostruct->outlier_direction_upwards[index] = 1;
         infostruct->base_yaxis[index] = 0;
      }
#endif
      switch ( j )
      {
         /* B ########################################################*/
         case 1: // 2nd version legend text; maybe (ikj)
            infostruct->legendtexts[index1] =
               bi_strdup( "Calculation Time in s (2)" ); // "... (ikj)"
            infostruct->legendtexts[index2] =
               bi_strdup( "FLOPS (2)" ); // "... (ikj)"
#ifdef BENCHIT_USE_PCL
            for ( i = 0; i < doable_nevents; i++ )
            {
               char buf[180];
               int index = n_of_works * ( n_of_sure_funcs_per_work + i ) + j;
               if ( doable_list[i] < PCL_MFLOPS )
                  sprintf( buf, "%160s per second (2)", PCLeventname( doable_list[i] ) );
               else
                  sprintf( buf, "%160s (2)", PCLeventname( doable_list[i] ) );
               infostruct->legendtexts[index] = bi_strdup( buf );
            }
#endif
            break;
         case 0: // 1st version legend text; maybe (ijk)
         default:
            infostruct->legendtexts[index1] =
               bi_strdup( "Calculation Time in s (1)" ); // "... (ijk)"
            infostruct->legendtexts[index2] =
               bi_strdup( "FLOPS (1)" ); // "... (ijk)"
#ifdef BENCHIT_USE_PCL
            for ( i = 0; i < doable_nevents; i++ )
            {
               char buf[180];
               int index = n_of_works * ( n_of_sure_funcs_per_work + i ) + j;
               if ( doable_list[i] < PCL_MFLOPS )
                  sprintf( buf, "%160s per second (1)", PCLeventname( doable_list[i] ) );
               else
                  sprintf( buf, "%160s (1)", PCLeventname( doable_list[i] ) );
               infostruct->legendtexts[index] = bi_strdup( buf );
            }
#endif
         /*########################################################*/
      }
   }
   if ( DEBUGLEVEL > 3 )
   {
      /* the next for loop: */
      /* this is for your information only and can be ereased if the kernel works fine */
      for ( i = 0; i < infostruct->numfunctions; i++ )
      {
         printf( "yaxis[%2d]=%s\t\t\tlegend[%2d]=%s\n",
            i, infostruct->yaxistexts[i], i, infostruct->legendtexts[i] );
      }
   }
}

/** Implementation of the bi_init of the BenchIT interface.
 *  Here you have the chance to allocate the memory you need.
 *  It is also possible to allocate the memory at the beginning
 *  of every single measurment and to free the memory thereafter.
 *  But making usage always of the same memory is faster.
 *  HAVE A LOOK INTO THE HOWTO !
 */
void* bi_init( int problemsizemax )
{
   mydata_t* mdp;
   mdp = (mydata_t*)malloc( sizeof( mydata_t ) );
   if ( mdp == 0 )
   {
      fprintf( stderr, "Allocation of structure mydata_t failed\n" ); fflush( stderr );
      exit( 127 );
   }
/*   if ( problemsizemax > STEPS )
   {
      fprintf( stderr, "Illegal maximum problem size\n" ); fflush( stderr );
      exit( 127 );
   }*/
   /* B ########################################################*/
   /* malloc our own arrays in here */
   /*########################################################*/
#ifdef BENCHIT_USE_PCL
   mdp->doable_list = doable_list;
   mdp->doable_nevents = doable_nevents;
   mdp->mode = mode;
   mdp->descr = descr;
   mdp->i_result = (PCL_CNT_TYPE*)malloc( doable_nevents * sizeof( PCL_CNT_TYPE ) );
   if ( mdp->i_result == 0 )
   {
      fprintf( stderr, "Allocation of i_result failed.\n" ); fflush( stderr );
      exit( 127 );
   }
   mdp->fp_result = (PCL_FP_CNT_TYPE*)malloc( doable_nevents * sizeof( PCL_FP_CNT_TYPE ) );
   if ( mdp->fp_result == 0 )
   {
      fprintf( stderr, "Allocation of fp_result failed.\n" ); fflush( stderr );
      exit( 127 );
   }
#endif
   return (void*)mdp;
}

/** The central function within each kernel. This function
 *  is called for each measurment step seperately.
 *  @param  mdpv         a pointer to the structure created in bi_init,
 *                       it is the pointer the bi_init returns
 *  @param  problemsize  the actual problemsize
 *  @param  results      a pointer to a field of doubles, the
 *                       size of the field depends on the number
 *                       of functions, there are #functions+1
 *                       doubles
 *  @return 0 if the measurment was sucessfull, something
 *          else in the case of an error
 */
int bi_entry( void* mdpv, int problemsize, double* results )
{
   /* ts, te: the start and end time of the measurement */
   /* timeinsecs: the time for a single measurement in seconds */
   double ts = 0.0, te = 0.0, timeinsecs = 0.0;
   /* flops stores the calculated FLOPS */
   double flops = 0.0;
   /* j is used for loop iterations */
   int j = 0;
   /* cast void* pointer */
   mydata_t* mdp = (mydata_t*)mdpv;

   /* calculate real problemsize */
   problemsize = MIN + ( problemsize - 1 ) * INCREMENT;

   /* check wether the pointer to store the results in is valid or not */
   if ( results == NULL ) return 1;

   /* B ########################################################*/
   /* maybe some init stuff in here */
   mdp->dummy = 0;
   /*########################################################*/

   for ( j = 0; j < n_of_works; j++ )
   {
      /* B ########################################################*/
      int index1 = 0 * n_of_works + j;
      int index2 = 1 * n_of_works + j;
      /* reset of reused values */
#ifdef BENCHIT_USE_PCL
      int i = 0;
      for ( i = 0; i < mdp->doable_nevents; i++ )
      {
         mdp->i_result[i] = 0;
         mdp->fp_result[i] = 0.0;
      }
#endif
      ts = 0.0;
      te = 0.0;
      /* choose version of algorithm */
      switch ( j ) {
         case 1: // 2nd version legend text; maybe (ikj)
            /* take start time, do measurment, and take end time */
#ifdef BENCHIT_USE_PCL
            if ( PCLstart( mdp->descr, mdp->doable_list,
               mdp->doable_nevents, mdp->mode ) != PCL_SUCCESS )
            {
               fprintf( stderr, "Problem with starting PCL events.\n" );
               bi_cleanup( mdpv ); fflush( stderr );
               exit( 1 );
            }
#endif
            ts = bi_timer(); work_2(); te = bi_timer();
#ifdef BENCHIT_USE_PCL
            if ( PCLstop( mdp->descr, mdp->i_result,
               mdp->fp_result, mdp->doable_nevents ) != PCL_SUCCESS )
            {
               fprintf( stderr, "Problem with stopping PCL events.\n" );
               bi_cleanup( mdpv ); fflush( stderr );
               exit( 1 );
            }
#endif
            break;
         case 0: // 1st version legend text; maybe (ijk)
         default:
            /* take start time, do measurment, and take end time */
#ifdef BENCHIT_USE_PCL
            if ( PCLstart( mdp->descr, mdp->doable_list,
               mdp->doable_nevents, mdp->mode ) != PCL_SUCCESS )
            {
               fprintf( stderr, "Problem with starting PCL events.\n" );
               bi_cleanup( mdpv ); fflush( stderr );
               exit( 1 );
            }
#endif
            ts = bi_timer(); work_1(); te = bi_timer();
#ifdef BENCHIT_USE_PCL
            if ( PCLstop( mdp->descr, mdp->i_result,
               mdp->fp_result, mdp->doable_nevents ) != PCL_SUCCESS )
            {
               fprintf( stderr, "Problem with stopping PCL events.\n" );
               bi_cleanup( mdpv ); fflush( stderr );
               exit( 1 );
            }
#endif
      }
      /* calculate the used time and FLOPS */
      timeinsecs = te - ts;
      timeinsecs -= dTimerOverhead;
      // this flops value is a made up! this calulations should be replaced
      // by something right for the choosen algorithm
      flops = problemsize;
      /* check for divide by zero, if timeinsecs is zero
      * timeinsecs=0 means that our operation does not need any time! */
      if ( timeinsecs < dTimerGranularity ) timeinsecs = INVALID_MEASUREMENT;
      /* store the results in results[1], results[2], ...
      * [1] for the first function, [2] for the second function
      * and so on ...
      * the index 0 always keeps the value for the x axis
      */
      /* B ########################################################*/
      // the xaxis value needs to be stored only once!
      if ( j == 0 ) results[0] = (double)problemsize;
      results[index1 + 1] = timeinsecs;
      results[index2 + 1] = flops;
#ifdef BENCHIT_USE_PCL
      for ( i = 0; i < mdp->doable_nevents; i++ )
      {
         int index = n_of_works * ( n_of_sure_funcs_per_work + i ) + j;
         if ( mdp->doable_list[i] < PCL_MFLOPS )
            results[index + 1] = (double)mdp->i_result[i] / ( 1.0 * timeinsecs );
         else
            results[index + 1] = mdp->fp_result[i];
      }
#endif
      /*########################################################*/
   }

   return ( 0 );
}

/** Clean up the memory
 */
void bi_cleanup( void* mdpv )
{
   mydata_t* mdp = (mydata_t*)mdpv;
   /* B ########################################################*/
   /* may be freeing our arrays here */
   /*########################################################*/
#ifdef BENCHIT_USE_PCL
   free(mdp->i_result);
   free(mdp->fp_result);
   free(mdp->doable_list);
   PCLexit(mdp->descr);
#endif
   if ( mdp ) free( mdp );
   return;
}
/********************************************************************/
/*************** End of interface implementations *******************/
/********************************************************************/

/* Reads the environment variables used by this kernel. */
void evaluate_environment()
{
   int errors = 0;
   char * p = 0;
   p = bi_getenv( "BENCHIT_KERNEL_PROBLEMSIZE_MIN", 0 );
   if ( p == 0 ) errors++;
   else MIN = atoi( p );
   p = bi_getenv( "BENCHIT_KERNEL_PROBLEMSIZE_MAX", 0 );
   if ( p == 0 ) errors++;
   else MAX = atoi( p );
   p = bi_getenv( "BENCHIT_KERNEL_PROBLEMSIZE_INCREMENT", 0 );
   if ( p == 0 ) errors++;
   else INCREMENT = atoi( p );
   if ( errors > 0 )
   {
      fprintf( stderr, "There's at least one environment variable not set!\n" );
      fprintf( stderr, "This kernel needs the following environment variables:\n" );
      fprintf( stderr, "BENCHIT_KERNEL_PROBLEMSIZE_MIN\n" );
      fprintf( stderr, "BENCHIT_KERNEL_PROBLEMSIZE_MAX\n" );
      fprintf( stderr, "BENCHIT_KERNEL_PROBLEMSIZE_INCREMENT\n" );
      fprintf( stderr, "\nThis kernel will iterate from BENCHIT_KERNEL_PROBLEMSIZE_MIN\n\
to BENCHIT_KERNEL_PROBLEMSIZE_MAX, incrementing by\n\
BENCHIT_KERNEL_PROBLEMSIZE_INCREMENT with each step.\n" );
      exit( 1 );
   }
}

/*****************************************************************************

LOG-History

$Log: kernel_main.c,v $
Revision 1.6  2005/12/13 00:40:33  mickler
+ Removed references to bi_info::log_[xy]axis
+ Changed for new unified min, max, increment format

Revision 1.5  2005/11/22 01:26:54  mickler
+ Using BI_GET_CALL_OVERHEAD_FUNC macro now
+ Using bi_timer() function for time measuring

Revision 1.4  2005/11/04 16:24:22  mickler
# Changed kernels to set base_[xy]axis=0 if not using logarithmic scale
- Fixed some kernels with log_[xy]axis=0 and base_[xy]axis=10 to base=0

Revision 1.3  2005/07/29 11:53:52  wloch
removed setting of kernelname in bi getinfo and equalized parameter variables

Revision 1.2  2005/07/20 12:48:52  wloch
kernellanguage and libraries are now taken from kernel name

Revision 1.1.1.1  2005/07/18 13:03:20  wloch
the final restructured benchit source tree

Revision 1.4  2005/06/21 12:27:24  wloch
adapted to new interface in JBI and benchit c

Revision 1.3  2005/06/16 15:16:40  wloch
adapted environment hashing changes to bi getenv

Revision 1.2  2005/02/15 11:01:23  wloch
spelling corrections

Revision 1.1  2005/02/11 10:38:20  wloch
code skeleton for c kernels using the PCL


*****************************************************************************/

File: benchit/Benchit-V5/kernel/numerical/matmul/Java/0/0/double/KernelMain.java
/********************************************************************
 * BenchIT - Performance Measurement for Scientific Applications
 *
 * Kernel: Matrix Multiply, Java
 * Contact: benchit@zih.tu-dresden.de
 *
 * Last change by: $Author: molka $
 * $Revision: 1.12 $
 * $Date: 2007/02/01 15:38:01 $
 *******************************************************************/

import java.util.Date;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;

/** This class is the Interface between the measured algorithm and BenchIT.
  * @author Robert Wloch, wloch@zhr.tu-dresden.de
  */
public class KernelMain implements BIJavaKernel {
   /* if it is static it has to be calculated just one time */
   private static double timeroverhead = 0.0;
   /**
    * Here we define a minimum time that our kernel needs
    * we do this to avoid a divide by zero.
    **/
   private static double MINTIME = 5.0e-6;

   /* attributes */
   /**
    * These variables will help us to keep the overview over the arrays
    * we access for our functions/data.
    **/
   /**
    * Number of different ways an algorithm will be measured.
    * Example: loop orders: ijk, ikj, jki, jik, kij, kji -> n_of_works=6 with
    * each different loop order in an own function.
    **/
   private int n_of_works;
   /**
    * Number of fixed functions we have per measurement.
    * Example: execution time and MFLOPS are measured for each loop order
    * -> n_of_sure_funcs_per_work=2
    **/
   private int n_of_sure_funcs_per_work;
   private int num_funcs = 0;
   private int bi_matmul_start, bi_matmul_stop, bi_matmul_increment;
   private BIEnvHash rwe = null;
   /**
    * The constructor.
    **/
   public KernelMain() {
      rwe = BIEnvHash.getInstance();
   }
   /**
    * The implementation of the bi_getinfo from the BenchIT interface.
    * Here the info is filled with informations about the kernel.
    * @param info An Object to hold the info about the kernel.
    **/
   public int bi_getinfo( InfoObject info ) {
      evaluate_environment();
      info.codesequence = "for( i=0; i<s; i++)#" +
                          "  for( j=0; j<s; j++)#" +
                          "    for( k=0; k<s; k++)#" +
                          "    {#" +
                          "      c[j*s+i]+=a[k*s+i]*b[j*s+k];#" +
                          "    }";
      info.xaxistext = "Matrix Size";
      info.log_xaxis = 0;
      info.base_xaxis = 0;
      info.maxproblemsize = (bi_matmul_stop-bi_matmul_start+1)/bi_matmul_increment;
      if ((bi_matmul_stop-bi_matmul_start+1) % bi_matmul_increment != 0) info.maxproblemsize++;
      info.kernel_execs_mpi1 = 0;
      info.kernel_execs_mpi2 = 0;
      info.kernel_execs_pvm = 0;
      info.kernel_execs_omp = 0;
      info.kernel_execs_pthreads = 0;
      info.kernel_execs_javathreads = 0;
      /* B ########################################################*/
      n_of_works = 6; /* number versions of this algorithm (ijk, ikj, kij, ... = 6 */
      n_of_sure_funcs_per_work = 1; /* FLOPS */
      /*########################################################*/
      num_funcs = n_of_works * n_of_sure_funcs_per_work;
      info.numfunctions = num_funcs;

      /* allocating memory for y axis texts and properties */
      info.yaxistexts = new String[info.numfunctions];
      info.outlier_direction_upwards = new int[info.numfunctions];
      info.legendtexts = new String[info.numfunctions];
      info.log_yaxis = new int[info.numfunctions];
      info.base_yaxis = new int[info.numfunctions];
      /* setting up y axis texts and properties */
      for ( int j = 0; j < n_of_works; j++ ) {
         /* B ########################################################*/
         int index1 = 0 * n_of_works + j;
         info.yaxistexts[index1] = "FLOPS";
         info.outlier_direction_upwards[index1] = 0;
         info.log_yaxis[index1] = 0;
         info.base_yaxis[index1] = 0;
         /*########################################################*/
         switch ( j ) {
            /* B ########################################################*/
            case 5: // 6th version legend text
               info.legendtexts[index1] = "FLOPS (kji)";
               break;
            case 4: // 5th version legend text
               info.legendtexts[index1] = "FLOPS (kij)";
               break;
            case 3: // 4th version legend text
               info.legendtexts[index1] = "FLOPS (jki)";
               break;
            case 2: // 3rd version legend text
               info.legendtexts[index1] = "FLOPS (jik)";
               break;
            case 1: // 2nd version legend text
               info.legendtexts[index1] = "FLOPS (ikj)";
               break;
            case 0: // 1st version legend text
            default:
               info.legendtexts[index1] = "FLOPS (ijk)";
            /*########################################################*/
         }
      }
      return 0;
   }
   /**
    * Implementation of the bi_init of the BenchIT interface.
    **/
   public Object bi_init( int problemsizemax ) {
      if ( problemsizemax > bi_matmul_stop ) {
         System.out.println( "Kernel: Illegal problem size!" );
         System.exit( 127 );
      }
      DataObject dataObject = new DataObject();
      return (Object)dataObject;
   }
   /**
    * The central function within each kernel. This function
    * is called for each measurment step seperately.
    * @param  dObject      an Object to the attributes initialized in bi_init,
    *                      it is the Object bi_init returns
    * @param  problemsize  the actual problemsize
    * @param  results      an array of doubles, the
    *                      size of the array depends on the number
    *                      of functions, there are #functions+1
    *                      doubles
    * @return 0 if the measurment was sucessfull, something
    *         else in the case of an error
    **/
   public int bi_entry( Object dObject, int problemsize, double[] results ) {
      /**
       * In java results is the total result array, not just a subarray
       * as in c. The offset points to the start part of the current entry.
       **/
      //int offset = ( num_funcs + 1 ) * ( problemsize - 1 );
      /* ts, te: the start and end time of the measurement */
      /* timeinsecs: the time for a single measurement in seconds */
      double ts = 0, te = 0;
      double timeinsecs = 0.0;
      /* flops stores the calculated FLOPS */
      double flops = 0.0;
      /* Cast Object reference */
      DataObject dataObject = (DataObject)dObject;
      dataObject.maxn = bi_matmul_start + (problemsize-1) * bi_matmul_increment;
      dataObject.a = new double[dataObject.maxn * dataObject.maxn];
      dataObject.b = new double[dataObject.maxn * dataObject.maxn];
      dataObject.c = new double[dataObject.maxn * dataObject.maxn];

      //System.out.println("problemsize: " + dataObject.maxn);

      /* calculate overhead if it is not done yet */
      if ( timeroverhead == 0.0 ) {
         timeroverhead = gettimeroverhead();
         /* maybe we have a VERY fast timer ;-) 
          * the /10000 is because we measure the timeroverhead with 10000 calls 
          * to the timer. if this can't be measured with the timers resolution
          * 1 call has to be 10000 times faster than the resolution of the timer
          */
         if ( timeroverhead == 0.0 ) timeroverhead = MINTIME/10000;
      }
      /* check wether the pointer to store the results in is valid or not */
      if ( results == null ) return 1;
      /* B ########################################################*/
      /* maybe some more init stuff in here */
      MatMul worker = new MatMul( dataObject );
      double flos = 2.0 * dataObject.maxn * dataObject.maxn * dataObject.maxn;
      worker.init();
      /*########################################################*/

      for ( int j = 0; j < n_of_works; j++ )
      {
         /* B ########################################################*/
         int index1 = 0 * n_of_works + j;
         /* reset of reused values */
         ts = 0;
         te = 0;
         timeinsecs = 0.0;
         /* choose version of algorithm */
         switch ( j ) {
            case 5: // 6th version legend text
               /* take start time, do measurment, and take end time */
               ts = JBI.bi_gettime();
               worker.matmul_kji();
               te = JBI.bi_gettime();
               break;
            case 4: // 5th version legend text
               /* take start time, do measurment, and take end time */
               ts = JBI.bi_gettime();
               worker.matmul_kij();
               te = JBI.bi_gettime();
               break;
            case 3: // 4th version legend text
               /* take start time, do measurment, and take end time */
               ts = JBI.bi_gettime();
               worker.matmul_jki();
               te = JBI.bi_gettime();
               break;
            case 2: // 3rd version legend text
               /* take start time, do measurment, and take end time */
               ts = JBI.bi_gettime();
               worker.matmul_jik();
               te = JBI.bi_gettime();
               break;
            case 1: // 2nd version legend text
               /* take start time, do measurment, and take end time */
               ts = JBI.bi_gettime();
               worker.matmul_ikj();
               te = JBI.bi_gettime();
               break;
            case 0: // 1st version legend text
            default:
               /* take start time, do measurment, and take end time */
               ts = JBI.bi_gettime();
               worker.matmul_ijk();
               te = JBI.bi_gettime();
         }
         /* calculate the used time and FLOPS */
         timeinsecs = te - ts; 
         timeinsecs -= timeroverhead;
         /* check for divide by zero, if timeinsecs is zero
         * timeinsecs=0 means that our operation does not need any time! */
         if ( timeinsecs < MINTIME ) timeinsecs = MINTIME;
         // this flops value is a made up! this calulations should be replaced
         // by something right for the choosen algorithm
         flops = (double)(flos / timeinsecs);
         /* store the results in results[1], results[2], ...
         * [1] for the first function, [2] for the second function
         * and so on ...
         * the index 0 always keeps the value for the x axis
         */
         /* B ########################################################*/
         // the xaxis value needs to be stored only once!
         if ( j == 0 ) results[0] = (double)dataObject.maxn;
         results[index1 + 1] = flops;
         /*########################################################*/
      }
      return(0);
   }
   /**
    * Tries to measure the timer overhead for a single call to bi_gettime().
    * @return the calculated overhead in seconds
    */
   private double gettimeroverhead() {
      int s;
      double ts, te, t;
      ts = JBI.bi_gettime();
      for ( s = 0; s < 10000; s++ ) {
         t = JBI.bi_gettime();
      }
      te = JBI.bi_gettime();
      return ( te - ts ) / 10000.0;
   }
   /**
    * Reads the environment variables used by this kernel.
    **/
   private void evaluate_environment() {
      String s_bi_matmul_start = null,  s_bi_matmul_stop = null, s_bi_matmul_increment = null;
      int i_bi_matmul_start = -1, i_bi_matmul_stop = -1, i_bi_matmul_increment = 500;
      s_bi_matmul_start = rwe.bi_getEnv("BENCHIT_KERNEL_PROBLEMSIZE_MIN");
      s_bi_matmul_stop = rwe.bi_getEnv("BENCHIT_KERNEL_PROBLEMSIZE_MAX");
      s_bi_matmul_increment = rwe.bi_getEnv("BENCHIT_KERNEL_PROBLEMSIZE_INCREMENT");
      if ((s_bi_matmul_start == null) || (s_bi_matmul_stop == null) || (s_bi_matmul_increment == null)) {
         i_bi_matmul_start = 1;
         i_bi_matmul_stop = 500;
         i_bi_matmul_increment = 1;
      }
      try {
         i_bi_matmul_start = (new Integer(s_bi_matmul_start)).intValue();
         i_bi_matmul_stop = (new Integer(s_bi_matmul_stop)).intValue();
         i_bi_matmul_increment = (new Integer(s_bi_matmul_increment)).intValue();
      }
      catch ( NumberFormatException nfe ) {
         i_bi_matmul_start = 1;
         i_bi_matmul_stop = 500;
         i_bi_matmul_increment = 1;
      }
      bi_matmul_start = i_bi_matmul_start;
      bi_matmul_stop = i_bi_matmul_stop;
      bi_matmul_increment = i_bi_matmul_increment;
   }
}


/********************************************************************
 * Log-History
 *
 *  $Log: KernelMain.java,v $
 *  Revision 1.12  2007/02/01 15:38:01  molka
 *  changed Java-Kernels to use C-Timer (JBI.bi_gettime()) instead of System.currentTimeMillis()
 *
 *  Revision 1.11  2006/05/11 11:15:03  hackenb
 *  still not perfect...
 *
 *  Revision 1.10  2006/05/10 20:41:22  hackenb
 *  implemented new iterating strategy
 *
 *  Revision 1.9  2006/05/09 11:15:11  rschoene
 *  axisbase is now int
 *
 *  Revision 1.8  2006/05/09 10:26:53  hackenb
 *  *** empty log message ***
 *
 *  Revision 1.7  2005/11/12 10:05:36  mickler
 *  - Fixed wrong subtraction (ts - te) in gettimeroverhead()
 *
 *  Revision 1.6  2005/11/04 16:24:22  mickler
 *  # Changed kernels to set base_[xy]axis=0 if not using logarithmic scale
 *  - Fixed some kernels with log_[xy]axis=0 and base_[xy]axis=10 to base=0
 *
 *  Revision 1.5  2005/09/11 22:56:28  mickler
 *  - Adapted MINTIME to timer granularity
 *  - Corrected names of env variables
 *
 *  Revision 1.4  2005/07/29 11:53:51  wloch
 *  removed setting of kernelname in bi getinfo and equalized parameter variables
 *
 *  Revision 1.3  2005/07/20 12:42:55  wloch
 *  kernellanguage and libraries are now taken from kernel name
 *
 *  Revision 1.2  2005/07/19 12:10:36  wloch
 *  added cvs footers
 * 
 *******************************************************************/
